### PR TITLE
[gosrc2cpg] - Receiver node or `thisParameterIn` node handling while processing method declaration 

### PR DESCRIPTION
**Pull request by pandurangpatil**
1. Changed the signature of `x2cpg` `NodeBuilders.newThisParameterNode` method to take `name` and `code` as a parameter and set defaults to `this`. This is done to use the receiver name as is defined in the code instead of naming it as `this`. This will make sure to have `ref` edges built as natural processing of `Identifiers.`  Have updated the places where this method was getting called to adjust with the updated interface.
2. `ThisParameterNode` handling while generating AST for method declaration that can be called with struct  type receiver

```
type Rect struct {
  len, wid int
}

func (re Rect) Area_by_value() int {
  return re.len * re.wid
}
```
3. Handling for more than one member, defined as a `,` separated list of the same type. 

```
package main

type Rect struct {
  len, wid int
}
```

### CODE REVIEW

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForFunctionsCreator.scala [hunk 1]
```diff
@@ -37,44 +37,91 @@ trait AstForFunctionsCreator(implicit withSchemaValidation: ValidationMode) { th
...
+    val (methodFullname, recSignStr) = receiverInfo match
+      case Some(_, typeFullName, evaluationStrategy, _) =>
+        val signatureStr = evaluationStrategy match
+          case EvaluationStrategies.BY_SHARING =>
+            s"*$typeFullName"
+          case EvaluationStrategies.BY_VALUE =>
+            typeFullName
+        (s"$typeFullName.$name", s"($signatureStr)")
+      case _ =>
+        (s"${fullyQualifiedPackage}.$name", "")
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForFunctionsCreator.scala:55]:
```suggestion
        (s"$fullyQualifiedPackage.$name", "")
```

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForFunctionsCreator.scala [hunk 2]
```diff
@@ -37,44 +37,91 @@ trait AstForFunctionsCreator(implicit withSchemaValidation: ValidationMode) { th
...
-      methodNode_.astParentType = NodeTypes.TYPE_DECL
-      Ast.storeInDiffGraph(astForMethod, diffGraph)
-      Seq.empty
-    } else {
-      Seq(astForMethod)
-    }
+    receiverInfo match
+      case Some(_, typeFullName, _, _) =>
+        // if method is related to Struct then fill astParentFullName and astParentType
+        methodNode_.astParentFullName = typeFullName
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForFunctionsCreator.scala:81]: Could we try to change these properties via setter methods for consistency?
```scala
methodNode_.astParentFullName(typeFullName)
    .astParentType(NodeTypes.TYPE_DECL)
```

File: joern-cli/frontends/jimple2cpg/src/main/scala/io/joern/jimple2cpg/passes/AstCreator.scala [hunk 3]
```diff
@@ -726,7 +726,11 @@ class AstCreator(filename: String, cls: SootClass, global: Global)(implicit with
             .argumentIndex(0)
             .dynamicTypeHintFullName(Seq(parentType))
         case _: NewMethodParameterIn =>
-          NodeBuilders.newThisParameterNode(parentType, Seq(parentType), line(Try(method.tryResolve()).getOrElse(null)))
+          NodeBuilders.newThisParameterNode(
+            typeFullName = parentType,
+            dynamicTypeHintFullName = Seq(parentType),
+            line = line(Try(method.tryResolve()).getOrElse(null))
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/jimple2cpg/src/main/scala/io/joern/jimple2cpg/passes/AstCreator.scala:732]: Is there not a way to have this `line` method accept `Option` types? Could go `Try(method.tryResolve()).toOption`

**Review by DavidBakerEffendi — COMMENTED** [review]
Minor style comments but otherwise LGTM, will accept once addressed

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForFunctionsCreator.scala (hunk 1 above)
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForFunctionsCreator.scala]: done

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForFunctionsCreator.scala (hunk 2 above)
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForFunctionsCreator.scala]: done

File: joern-cli/frontends/jimple2cpg/src/main/scala/io/joern/jimple2cpg/passes/AstCreator.scala (hunk 3 above)
Comment (pandurangpatil) [reply joern-cli/frontends/jimple2cpg/src/main/scala/io/joern/jimple2cpg/passes/AstCreator.scala:732]: `line(Try(method.tryResolve()).getOrElse(null))` method which is getting called returns and `Option` check line no `1148`

Comment (DavidBakerEffendi) [reply joern-cli/frontends/jimple2cpg/src/main/scala/io/joern/jimple2cpg/passes/AstCreator.scala:732]: Ok could we try
```scala
line = Try(method.tryResolve()).toOption.map(line)
```

Comment (pandurangpatil) [reply joern-cli/frontends/jimple2cpg/src/main/scala/io/joern/jimple2cpg/passes/AstCreator.scala:732]: line() expects a `Host` object in this map it passes `Option`. Hence it throws error.

I have just updated the method call to use named parameters with the change I made in the method parameters.

image

**Comment by pandurangpatil** [comment]
@DavidBakerEffendi I pushed one more fix for another ignored Unit test

**Review by DavidBakerEffendi — APPROVED** [review]

**Pull request merged**